### ISSUE TITLE
new lint for &str to String conversion

### ISSUE THREAD

**Issue by mockersf** [rust-lang-nursery/rust-clippy#2824]
There are a few different ways to convert from a `&str` to a `String`:
```rust
fn main() {
    let _: String = "hello".to_owned();
    let _: String = "hello".to_string();
    let _: String = "hello".into();
    let _: String = String::from("hello");
    let _: String = format!("hello");
}
```

Without going into a discussion about which is the best ([this lint](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#str_to_string) has been deprecated), I think a lint to ensure the same one is used throughout a crate would be interesting.

This could be done either through configuration, or by keeping stats on which is used and then emit warning for the non majority one.

I can work on it, but wanted to have a general point of view before. What do you think ?

**Comment by oli-obk** [comment]
I think we can rule out the `format!` variant, it's very expensive. And we can rule out `into` as that only works if the target type is specified (you can't call `"hello".into().len()`).

So we're left with

* `"hello".to_owned()`
* `"hello".to_string()`
* `String::from("hello")`

Statistics on the crate seem like a good idea, but might end up being expensive (you need to first go through the entire crate looking at all method/function calls, collect the stats, and then run the lint normally). So we'll need to benchmark some large crates and see how expensive it is in reality.

**Comment by clarcharr** [comment]
Honestly, I'd rule out `String::from` for literals, as the method style is more idiomatic.

**Comment by oli-obk** [comment]
I remember @steveklabnik having opinions on this

**Comment by steveklabnik** [comment]
The official style is to use `.to_string()` everywhere. There's also a pretty significant minority style of "`String::from` for literals, `.to_string()` for variables." IMHO, `to_owned` is for generic code only; you should say directly what you mean: I want a `String`.

(`format!` is, IIRC, significantly slower; that's why people started using `.to_owned`, because it skipped the formatting machinery, whereas `.to_string` did not. Now they're equivalent.)

**Comment by spunit262** [comment]
I have to disagree. `"abc".to_string()` fails to describe what it's doing, it only says it's making a string form a string. `"abc".to_owned()` however highlight the important part of what it's doing, creating an *owned* string.

It's not much of a difference, but `to_owned` is slightly more newcomer friendly than `to_string` and as such should generally be preferred.

**Comment by Kixunil** [comment]
@spunit262 exactly, `to_string()` is meant to *display* stuff. When newbie sees:

```rust
lex foo = "foo".to_string();
```

"WTF, isn't it a string already?"

It's actually the other way around: use `to_string()` in generic code if you need to display stuff.
